### DISCUSSION TITLE
Prohibit paste on specific columns/cells

### DISCUSSION THREAD

**Discussion by zewa666** [ghiscoding/Angular-Slickgrid#1339] [General]
So I've been looking around stackoverflow and the code base but didn't find anything in order to be able to limit the paste action both for the Excel Copy Buffer and normal copy pastes.

Is there something like a onBeforeCellPaste event, in tandem with onPasteCells where I could return a boolean to decide whether the paste should happen or not?

Alternatively, it might also be an option to modify the paste behavior to not allow pastes if the target cell/col has no editor assigned.

I've tried using onBeforeEditCell of the column, but that does not get triggered for pastes. Using that would be a third option as far as I see.

If there is no way yet existing to support this functionally, as always, I'm happy to provide a PR upon agreement of which path you'd like to approach @ghiscoding 

EDIT: I recon this would be primarily a PR for the Slickgrid-Universal repo, sorry for starting the discussion here. I just thought about it after writing this :)

**Comment by ghiscoding** [comment]
I think there was similar discussion on the SlickGrid project but I don't think anything came out of that. The thing is that these plugins were created by other users long time ago, the only thing I've done was to migrate them to the project and add unit tests for what existed. So yes feel free to create a PR, below are the steps you can probably take

~I don't see any SlickEvent in the [`slickCellExcelCopyManager.ts`](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/common/src/extensions/slickCellExcelCopyManager.ts)~ ah nevermind there's actually in the [slickCellExternalCopyManager.ts](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/common/src/extensions/slickCellExternalCopyManager.ts) file, so I guess you could add some new events for your use case in there. I see an event called `onCopyCancelled` [here](https://github.com/ghiscoding/slickgrid-universal/blob/9d60a9d82e605ae2351822c66ff8757349b906cf/packages/common/src/extensions/slickCellExternalCopyManager.ts#L352-L361) but it's only for the `Escape` key, I'm not sure if you want to go with that event and simply create a new one, whichever you think is best. 

The approach that we used for cancelling (e.g. `onBeforeEditCell`) is that by calling the SlickEvent `notify` that will return an object which you can then call `.getReturnValue()` on it and compare it against `false` and if so then simply `return`, for example I've done that recently on the Row Detail cancel opening on the SlickGrid project [PR](https://github.com/6pac/SlickGrid/pull/852/files)
```ts
// user could cancel the Row Detail opening when event is returning false
if (this.onBeforeRowDetailToggle.notify({ grid: this._grid, item: dataContext }, e, this).getReturnValue() === false) {
  return;
}
```
That's great but how will you have access to that code on the user's side? So in order to expose these stuff in the grid option, I simply subscribe to the SlickEvent and when it's called, I then compare with an if statement and execute the code when it's defined in the grid option, for example with the `onCopyCancelled` [here](https://github.com/ghiscoding/slickgrid-universal/blob/master/packages/common/src/extensions/slickCellExcelCopyManager.ts#L77-L81)

Oh and I created a list of all the events in the docs [here](https://ghiscoding.gitbook.io/angular-slickgrid/events/available-events) that exist in the project but for some reasons it looks like I've missed the one from the `slickCellExternalCopyManager.ts`, we should probably add them to the docs, the docs are now part of the code and under `docs/events/available-events`

> EDIT: I recon this would be primarily a PR for the Slickgrid-Universal repo, sorry for starting the discussion here. I just thought about it after writing this :)

Yeah it would have been better since I can't copy permalink of code to another repo :D

**Reply by zewa666** [reply]
oh wow I missed this column option
https://github.com/ghiscoding/slickgrid-universal/blob/9d60a9d82e605ae2351822c66ff8757349b906cf/packages/common/src/extensions/slickCellExternalCopyManager.ts#L125

perhaps that solves it already. will give it a try and if so add some docs about. good to see you switched to gitbook. way easier than keeping the wiki up to date

**Reply by zewa666** [reply]
alright this works perfectly, even in multi col spanning table buffer pastes. I've created a docs PR for that
